cpp/src/parquet/file_writer.cc

```diff
@@ -567,6 +567,44 @@ void WriteEncryptedFileMetadata(const FileMetaData& file_metadata,
   }
 }
 
+void WriteEncryptedMetadataFile(
+    const FileMetaData& metadata, std::shared_ptr<::arrow::io::OutputStream> sink,
+    std::shared_ptr<FileEncryptionProperties> file_encryption_properties) {
+  auto file_encryptor = std::make_unique<InternalFileEncryptor>(
+      file_encryption_properties.get(), ::arrow::default_memory_pool());
+
+  if (file_encryption_properties->encrypted_footer()) {
+    PARQUET_THROW_NOT_OK(sink->Write(kParquetEMagic, 4));
+
+    PARQUET_ASSIGN_OR_THROW(int64_t position, sink->Tell());
+    auto metadata_start = static_cast<uint64_t>(position);
+
+    auto writer_props = parquet::WriterProperties::Builder()
+                            .encryption(file_encryption_properties)
+                            ->build();
+    auto builder = FileMetaDataBuilder::Make(metadata.schema(), writer_props);
+
+    auto footer_metadata = builder->Finish(metadata.key_value_metadata());
+    auto crypto_metadata = builder->GetCryptoMetaData();
+    WriteFileCryptoMetaData(*crypto_metadata, sink.get());
+
+    auto footer_encryptor = file_encryptor->GetFooterEncryptor();
+    WriteEncryptedFileMetadata(metadata, sink.get(), footer_encryptor, true);
+    PARQUET_ASSIGN_OR_THROW(position, sink->Tell());
+    auto footer_and_crypto_len = static_cast<uint32_t>(position - metadata_start);
+    PARQUET_THROW_NOT_OK(
+        sink->Write(reinterpret_cast<uint8_t*>(&footer_and_crypto_len), 4));
+    PARQUET_THROW_NOT_OK(sink->Write(kParquetEMagic, 4));
+  } else {
```

Comment:
I'm doing metadata encryption similarly to approach here: https://github.com/apache/arrow/blob/a4d58e0c8213cd14d0953c75eb047ea1693fddc9/cpp/src/parquet/file_writer.cc#L416-L432
However it seems decryption fails (see below) when using `RowGroup` metadata (after deserialization and decryption).
https://github.com/apache/arrow/blob/4d5041a3eff486cd82c6790f6dae63fba2321474/cpp/src/arrow/dataset/file_parquet.cc#L1084-L1086 This makes me think I'm either not serializing correctly or there's an issue with encryption/decryption properties I'm supplying.

@wgtmac @pitrou does anything obviously wrong stands out here?

Reply:
cc @mapleFU

Reply:
What error did you get? I suspect it has some issues with row group ordinal. Please search "row group ordinal" from https://github.com/apache/parquet-format/blob/master/Encryption.md

Reply:
The actual error is thrown here: https://github.com/apache/arrow/blob/4d5041a3eff486cd82c6790f6dae63fba2321474/cpp/src/arrow/dataset/file_parquet.cc#L1009
in debugger I see `row_group.ColumnChunk(0)` as null and `row_group.ColumnChunk(0)->file_path()` fails.

Reply:
For the test `FileMetadata` object this reads several files, extracts their metadata and merges them (`metadata->AppendRowGroups(*metadata_vector[1])`).

Reply:
We could solve this by storing file aad prefixes into `key_value_metadata` so they can be used at read time to decode row group metadata. This seems to work (see [cpp/src/parquet/metadata.cc](https://github.com/apache/arrow/pull/41821/files#diff-a5c0680e0cdc891095924b93ac23383d3a962b7b929876818ecffe461f83b1cd)), but reading actual data with `scanner->ToTable()` is somewhat less straightforward. but seems doable. Before proceeding, I'd like to ask here:
- does this approach make sense?
- what would a good location be for injecting aad prefixes be for reading column data?

cc @wgtmac @ggershinsky

Reply:
The format has a field for aad prefixes, https://github.com/apache/parquet-format/blob/master/src/main/thrift/parquet.thrift#L1126 .

It is file-wide, of course. So if parquet files had different aad prefixes, a new one has to be used for the merged metadata. 

Per my comment above, I think the only way to make this work is to fully decrypt the footers of the parquet files (using their encryption params, inc keys and aad_prefixes), merge them, and then encrypt the result with a new key/aad_prefix